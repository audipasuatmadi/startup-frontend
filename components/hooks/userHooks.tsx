import useSWR from 'swr';
import { getFetcher } from '../apis/MainFetchers';
import { AxiosError } from 'axios';
import { fetchFromStorage } from '../apis/localstorage_apis/LocalStorageApi';

export interface UserData {
  name: string;
  username: string;
  profileURL?: string;
}

export interface UseUserDataReturn {
  userData?: UserData,
  isLoading: boolean,
  isError?: AxiosError
}

export interface FullUserData {
  id: number;
  name: string;
  username: string;
  profileURL?: string;
  token: string;
  createdAt: string;
  updatedAt: string;
}

export const useUserData = (id: string) => {
  const { data, error } = useSWR<UserData | undefined, AxiosError>(
    `users/${id}`,
    getFetcher
  );

  return {
    userData: data,
    isLoading: !error && !data,
    isError: error,
  };
};

export const useUserStorage = () => {
  const { data } = useSWR<FullUserData | null>('user', fetchFromStorage)
  return data
}