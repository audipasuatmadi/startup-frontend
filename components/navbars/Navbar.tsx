import React, { useState, useEffect, useReducer } from 'react';
import Link from 'next/link';
import Image from 'next/image';
import TextField from '../inputfields/TextField';
import NavigationItem from './NavigationItem';
import Button from '../buttons/Button';
import IconButton from '../buttons/IconButton';
import { useSelector, useDispatch } from 'react-redux';
import { RootState } from '../../states/RootReducer';
import { isUserDataType, UserData } from '../../lib/user/userTypes';
import Dropdown from '../dropdowns/Dropdown';
import DropdownItem from '../dropdowns/DropdownItem';
import { getLocalStorageData } from '../../lib/utils/LocalStorageUtil';

const searchIcon = (
  <Image
    src='/assets/svgs/ic-search.svg'
    width={25}
    height={25}
    alt='search bar image'
  />
);
const hamburgerIcon = (
  <Image
    src='/assets/svgs/ic-hamburger-menu.svg'
    width={25}
    height={25}
    alt='search bar image'
  />
);

const ProfileDropdown = ({userData}: {userData: UserData}) => (
  <Dropdown
    togglerRender={(props) => (
      <IconButton
        className="w-7 h-7 md:w-10 md:h-10"
        {...props}
        icon={
          <div
            className='h-full w-full rounded-full bg-gray-700'
          >
          </div>
        }></IconButton>
    )}
    containerAlign='right'>
    <DropdownItem disabled>{userData.name}</DropdownItem>
    <DropdownItem divider>Profil Saya</DropdownItem>
    <DropdownItem>Bookmarks</DropdownItem>
    <DropdownItem divider>Pengaturan</DropdownItem>
    <DropdownItem>Privasi</DropdownItem>
    <DropdownItem>Logout</DropdownItem>
  </Dropdown>
);

const Navbar = () => {
  const [isMenuOpened, toggleIsMenuOpened] = useReducer(
    (isMenuOpened) => !isMenuOpened,
    false
  );

  const dispatch = useDispatch();

  useEffect(() => {
    const accessToken = getLocalStorageData('at');
    const refreshToken = getLocalStorageData('rt');

    if (refreshToken && accessToken) {
      
    }
  }, [])

  const userDataState = useSelector((state: RootState) => state.userData);

  return (
    <nav className='fixed w-screen flex items-center py-2 shadow h-14 bg-white z-50'>
      <Link href='/'>
        <a className='ml-8 whitespace-nowrap text-brand font-bold'>
          Elites Bible
        </a>
      </Link>

      <TextField
        id='searchBar'
        name='searchBar'
        type='text'
        placeholder='Pencarian...'
        solid
        leadingIcon={searchIcon}
        className='hidden md:inline-block md:w-3/12 lg:w-5/12 ml-12 mr-2'
      />

      <ul className='ml-auto md:hidden mr-5 whitespace-nowrap flex items-center'>
        <li className='sm:space-y-0 md:space-y-0 '>
          <IconButton icon={searchIcon} />
        </li>
        {
          isUserDataType(userDataState) && (
            <ProfileDropdown userData={userDataState} />
          )
        }
        <li>
          <IconButton onClick={toggleIsMenuOpened} icon={hamburgerIcon} />
        </li>
      </ul>

      <div
        className={`fixed md:static w-screen h-screen top-0 mt-14 md:mt-0 md:top-auto md:w-auto md:h-auto sm:ml-auto bg-gray-200 md:bg-transparent ${
          isMenuOpened ? 'flex' : 'hidden md:flex'
        } justify-center items-center`}>
        <ul className='flex flex-col space-y-8 md:space-y-0 w-max md:ml-auto md:flex-row md:space-x-10 md:mr-14 items-center'>
          <NavigationItem href='/'>Beranda</NavigationItem>
          <NavigationItem href='/trending'>Trending</NavigationItem>
          <NavigationItem href='/articles'>Artikel</NavigationItem>
          <NavigationItem href='/courses'>Courses</NavigationItem>
          <NavigationItem>
            {isUserDataType(userDataState) ? (
              <ProfileDropdown userData={userDataState}   />
            ) : (
              <Button rounded href='/login'>
                Bergabung
              </Button>
            )}
          </NavigationItem>
        </ul>
      </div>
    </nav>
  );
};

export default Navbar;
