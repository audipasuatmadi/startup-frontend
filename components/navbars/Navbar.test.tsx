import React from 'react';
import { shallow, ShallowWrapper, mount } from 'enzyme';
import Navbar from './Navbar';
import { useUserData, UseUserDataReturn } from '../hooks/userHooks';

jest.mock('../hooks/userHooks.tsx');

const dummyLoggedInData: UseUserDataReturn = {
  isLoading: false,
  userData: {
    name: 'John Doe',
    username: 'johndoe',
  },
};

const dummyLoggedOutData: UseUserDataReturn = {
  isLoading: false,
};

const dummyIsLoadingData: UseUserDataReturn = {
  isLoading: true,
};

describe('navbar functionality tests', () => {
  let mockedUseUserData: jest.Mock<UseUserDataReturn>;

  mockedUseUserData = useUserData as jest.Mock<UseUserDataReturn>;

  afterEach(() => {
    mockedUseUserData.mockClear();
  });

  it('should call the useUserData hook properly', () => {
    mockedUseUserData.mockReturnValue({ isLoading: false });

    shallow(<Navbar />);

    expect(useUserData).toBeCalled();
    expect(useUserData).toHaveReturnedWith<UseUserDataReturn>({
      isLoading: false,
    });
  });

  it('should have the right menu for logged out user', () => {
    mockedUseUserData.mockReturnValue(dummyLoggedOutData);

    const wrapper = shallow(<Navbar />);

    expect(useUserData).toHaveReturnedWith<UseUserDataReturn>(
      dummyLoggedOutData
    );
    expect(wrapper.find('Button').length).toBe(3);
    expect(wrapper.find('Menu').length).toBe(1);
  });

  it('should have the right menu on loading user data', () => {
    mockedUseUserData.mockReturnValue(dummyIsLoadingData)

    const wrapper = shallow(<Navbar />);

    expect(useUserData).toHaveReturnedWith<UseUserDataReturn>(dummyIsLoadingData)
    expect(wrapper.find('Button').length).toBe(3)
    expect(wrapper.find('Button').last().prop('disabled')).toEqual(true)
  })

  it('should have the right menu for logged in user', () => {
    mockedUseUserData.mockReturnValue(dummyLoggedInData)

    const wrapper = shallow(<Navbar />);

    expect(useUserData).toHaveReturnedWith<UseUserDataReturn>(dummyLoggedInData)
    expect(wrapper.find('Button').length).toBe(2)
    expect(wrapper.find('Menu').length).toBe(2)

  })
});
