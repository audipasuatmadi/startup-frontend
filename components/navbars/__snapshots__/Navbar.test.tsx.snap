// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar rendering tests should render the main components correctly 1`] = `
<nav
  className="fixed w-screen flex items-center py-2 shadow h-14 bg-white z-50"
>
  <Link
    href="/"
  >
    <a
      className="ml-8 whitespace-nowrap text-brand font-bold"
    >
      Elites Bible
    </a>
  </Link>
  <TextField
    className="hidden md:inline-block md:w-3/12 lg:w-5/12 ml-12 mr-2"
    id="searchBar"
    leadingIcon={
      <Image
        alt="search bar image"
        height={25}
        src="/assets/svgs/ic-search.svg"
        width={25}
      />
    }
    name="searchBar"
    placeholder="Pencarian..."
    solid={true}
    type="text"
  />
  <ul
    className="ml-auto md:hidden mr-5 whitespace-nowrap flex items-center"
  >
    <li
      className="sm:space-y-0 md:space-y-0 "
    >
      <IconButton
        icon={
          <Image
            alt="search bar image"
            height={25}
            src="/assets/svgs/ic-search.svg"
            width={25}
          />
        }
      />
    </li>
    <li>
      <IconButton
        icon={
          <Image
            alt="search bar image"
            height={25}
            src="/assets/svgs/ic-hamburger-menu.svg"
            width={25}
          />
        }
        onClick={[Function]}
      />
    </li>
  </ul>
  <div
    className="fixed md:static w-screen h-screen top-0 mt-14 md:mt-0 md:top-auto md:w-auto md:h-auto sm:ml-auto bg-gray-200 md:bg-transparent hidden md:flex justify-center items-center"
  >
    <ul
      className="flex flex-col space-y-8 md:space-y-0 w-max md:ml-auto md:flex-row md:space-x-10 md:mr-14 items-center"
    >
      <NavigationItem
        href="/"
      >
        Beranda
      </NavigationItem>
      <NavigationItem
        href="/trending"
      >
        Trending
      </NavigationItem>
      <NavigationItem
        href="/articles"
      >
        Artikel
      </NavigationItem>
      <NavigationItem
        href="/courses"
      >
        Courses
      </NavigationItem>
      <NavigationItem>
        <Button
          href="/login"
          rounded={true}
          theme="brand"
        >
          Bergabung
        </Button>
      </NavigationItem>
    </ul>
  </div>
</nav>
`;

exports[`Navbar rendering tests should render the right buttons when user is not logged out 1`] = `
<nav
  className="fixed w-screen flex items-center py-2 shadow h-14 bg-white z-50"
>
  <Link
    href="/"
  >
    <a
      className="ml-8 whitespace-nowrap text-brand font-bold"
    >
      Elites Bible
    </a>
  </Link>
  <TextField
    className="hidden md:inline-block md:w-3/12 lg:w-5/12 ml-12 mr-2"
    id="searchBar"
    leadingIcon={
      <Image
        alt="search bar image"
        height={25}
        src="/assets/svgs/ic-search.svg"
        width={25}
      />
    }
    name="searchBar"
    placeholder="Pencarian..."
    solid={true}
    type="text"
  />
  <ul
    className="ml-auto md:hidden mr-5 whitespace-nowrap flex items-center"
  >
    <li
      className="sm:space-y-0 md:space-y-0 "
    >
      <IconButton
        icon={
          <Image
            alt="search bar image"
            height={25}
            src="/assets/svgs/ic-search.svg"
            width={25}
          />
        }
      />
    </li>
    <li>
      <IconButton
        icon={
          <Image
            alt="search bar image"
            height={25}
            src="/assets/svgs/ic-hamburger-menu.svg"
            width={25}
          />
        }
        onClick={[Function]}
      />
    </li>
  </ul>
  <div
    className="fixed md:static w-screen h-screen top-0 mt-14 md:mt-0 md:top-auto md:w-auto md:h-auto sm:ml-auto bg-gray-200 md:bg-transparent hidden md:flex justify-center items-center"
  >
    <ul
      className="flex flex-col space-y-8 md:space-y-0 w-max md:ml-auto md:flex-row md:space-x-10 md:mr-14 items-center"
    >
      <NavigationItem
        href="/"
      >
        Beranda
      </NavigationItem>
      <NavigationItem
        href="/trending"
      >
        Trending
      </NavigationItem>
      <NavigationItem
        href="/articles"
      >
        Artikel
      </NavigationItem>
      <NavigationItem
        href="/courses"
      >
        Courses
      </NavigationItem>
      <NavigationItem>
        <Button
          href="/login"
          rounded={true}
          theme="brand"
        >
          Bergabung
        </Button>
      </NavigationItem>
    </ul>
  </div>
</nav>
`;
